### PR TITLE
support both tag and digest for pulling images

### PR DESCRIPTION
**Pull request by zhangguanzhang**
Fixes: https://github.com/containers/podman/issues/6721
pull with both a tag and digest
```
$ docker pull docker.io/fedora:32@sha256:e69b5a62ce23c673885bddc94e6679c9b2af683059637ceddb9cff458537a326
$ docker images | grep fedora
fedora              <none>              2a8d315e6208        7 months ago        201MB
$ docker rmi 2a8d
```
pull without a tag
```
$ docker pull docker.io/fedora@sha256:e69b5a62ce23c673885bddc94e6679c9b2af683059637ceddb9cff458537a326
$ docker images | grep fedora
fedora              <none>              2a8d315e6208        7 months ago        201MB
```

pull with a wrong tag with the digest
```
$ docker pull docker.io/fedora:31@sha256:e69b5a62ce23c673885bddc94e6679c9b2af683059637ceddb9cff458537a326
$ docker images | grep fedora
fedora              <none>              2a8d315e6208        7 months ago        201MB
```
so the tag will be ignore when with the digest

Signed-off-by: zhangguanzhang <zhangguanzhang@qq.com>

### CODE REVIEW

**Comment by zhangguanzhang** [comment]
@mheon PATL
```
  podman pull with both a tag and digest
           /var/tmp/go/src/github.com/containers/podman/test/e2e/pull_test.go:94
         
         [BeforeEach] Podman pull
           /var/tmp/go/src/github.com/containers/podman/test/e2e/pull_test.go:22
         [It] podman pull with both a tag and digest
           /var/tmp/go/src/github.com/containers/podman/test/e2e/pull_test.go:94
         Running: podman [options] pull quay.io/libpod/alpine:3.10.2@sha256:fa93b01658e3a5a1686dc3ae55f170d8de487006fb53a28efcd12ab0710a2e5f
         Trying to pull quay.io/libpod/alpine@sha256:fa93b01658e3a5a1686dc3ae55f170d8de487006fb53a28efcd12ab0710a2e5f...
         Getting image source signatures
         Copying blob sha256:9d16cba9fb961d1aafec9542f2bf7cb64acfc55245f9e4eb5abecd4cdc38d749
         Copying config sha256:961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4
         Writing manifest to image destination
         Storing signatures
         961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4
         output: 961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4 <--- id
         Running: podman [options] rmi 961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4 <--- but failed
         Error: 1 error occurred:
         	* layer not known
```

**Comment by rhatdan** [comment]
@mtrmac @vrothberg PTAL
@containers/podman-maintainers PTAL

File: cmd/podman/images/pull.go [hunk 1]
```diff
@@ -156,9 +156,21 @@ func imagePull(cmd *cobra.Command, args []string) error {
 		pullOptions.Username = creds.Username
 		pullOptions.Password = creds.Password
 	}
+
+	rawImgName := args[0]
```
Comment (mtrmac) [inline cmd/podman/images/pull.go:160]: This looks like the wrong place. What about `podman run` and other places that auto-pull? Non-CLI users? The Docker HTTP API?

Also, an _intentional decision_ about whether such syntax is allowed for already-pulled images needs to happen.

File: cmd/podman/images/pull.go [hunk 2]
```diff
@@ -156,9 +156,21 @@ func imagePull(cmd *cobra.Command, args []string) error {
...
+	// TODO: maybe need update at github.com/containers/image/v5/docker/docker_transport.go:79-81
```
Comment (mtrmac) [inline cmd/podman/images/pull.go:161]: No. See earlier discussion.

File: cmd/podman/images/pull.go [hunk 3]
```diff
@@ -156,9 +156,21 @@ func imagePull(cmd *cobra.Command, args []string) error {
...
+	// if use tag and digest, the tag should be ignored
+	if !strings.Contains(rawImgName, "://") {
```
Comment (mtrmac) [inline cmd/podman/images/pull.go:163]: This is a loose, and _undocumented_, heuristic. What is the exact condition this is trying to implement? This is wrong for the two conditions I can think of.

File: cmd/podman/images/pull.go [hunk 4]
```diff
@@ -156,9 +156,21 @@ func imagePull(cmd *cobra.Command, args []string) error {
...
+		rawImgParts := strings.Split(rawImgName, ":")
+		if len(rawImgParts) == 3 && strings.Contains(rawImgParts[1], "@sha256") {
```
Comment (mtrmac) [inline cmd/podman/images/pull.go:165]: There are proper parsing libraries for references. Please don’t add heuristic implementations like these that make hard-to-predict edits on invalid data.

**Review by mtrmac — CHANGES_REQUESTED** [review]

File: cmd/podman/images/pull.go (hunk 1 above)
Comment (zhangguanzhang) [reply cmd/podman/images/pull.go:160]: Do I need to modify it in  `containers/image`

Comment (mtrmac) [reply cmd/podman/images/pull.go:160]: Please read all of https://github.com/containers/podman/issues/6721 including the linked conversations.

**Pull request closed without merging**